File: Trees/230_Kth_Smallest_Element_in_BST/solution230.cpp
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> stk;
        TreeNode* curr = root;

        while (!stk.empty() || curr) {
            while (curr) {
                stk.push(curr);
                curr = curr->left;
            }
            
            curr = stk.top();
            stk.pop();
            
            k--;
            if (k == 0) {
                return curr->val;
            }
            
            curr = curr->right;
        }
        
        return -1;
    }
};

// Runtime1: 11ms, beats 72.91%
// Memory1: 22.99MB, beats 38.94%

// Runtime2: 7ms, beats 93.98%
// Memory2: 22.90MB, beats 38.94%

// Runtime3: 11ms, beats 72.91%
// Memory3: 22.78MB, beats 89.16%

File: Trees/230_Kth_Smallest_Element_in_BST/solution230.py
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        stack = []
        curr = root

        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            k -= 1
            if k == 0:
                return curr.val
            curr = curr.right
            
# Runtime1: 40ms, beats 83.55%
# Memory1: 19.40MB, beats 97.80%

# Runtime2: 33ms, beats 97.90%
# Memory2: 19.48MB, beats 76.97%

# Runtime3: 35ms, beats 96.12%
# Memory3: 19.48MB, beats 76.97%